File: leetcode/remove-duplicates-from-sorted-array-ii.py
# 80. Remove Duplicates from Sorted Array II
# 🟠 Medium
#
# https://leetcode.com/problems/remove-duplicates-from-sorted-array-ii/
#
# Tags: Array - Two Pointers

import timeit
from typing import List


# Remove duplicates leaving one or two instances of every unique value.
# Iterate over the input keeping the count of times we have seen the
# last value, when the count is under one, write the value under the
# read pointer on the position indicated by the write pointer, when we
# have seen the value under the read pointer twice already, move the
# read pointer forward without moving the write pointer.
#
# Time complexity: O(n) - We visit each value in the input once.
# Space complexity: O(1) - We use constant memory.
#
# Runtime: 60 ms, faster than 88.76%
# Memory Usage: 13.8 MB, less than 74.73%
class TwoPointers:
    def removeDuplicates(self, nums: List[int]) -> int:
        w = count = 1
        for r in range(1, len(nums)):
            if nums[r] != nums[w - 1]:
                nums[w], count = nums[r], 1
                w += 1
            elif count < 2:
                count += 1
                nums[w] = nums[r]
                w += 1
        return w


def test():
    executors = [
        TwoPointers,
    ]
    tests = [
        [[1, 1, 1, 2, 2, 3], 5],
        [[0, 0, 1, 1, 1, 1, 2, 3, 3], 7],
    ]
    for executor in executors:
        start = timeit.default_timer()
        for _ in range(1):
            for col, t in enumerate(tests):
                sol = executor()
                result = sol.removeDuplicates(t[0])
                exp = t[1]
                assert result == exp, (
                    f"\033[93m» {result} <> {exp}\033[91m for"
                    + f" test {col} using \033[1m{executor.__name__}"
                )
        stop = timeit.default_timer()
        used = str(round(stop - start, 5))
        cols = "{0:20}{1:10}{2:10}"
        res = cols.format(executor.__name__, used, "seconds")
        print(f"\033[92m» {res}\033[0m")


test()

File: leetcode/remove-duplicates-from-sorted-array-ii.rs
// 80. Remove Duplicates from Sorted Array II
// 🟠 Medium
//
// https://leetcode.com/problems/remove-duplicates-from-sorted-array-ii/h
//
// Tags: Array - Two Pointers

struct Solution;
impl Solution {
    // Generic solution that leaves a maximum of n instances of the same element
    // in a given sorted input array. The solution uses two pointers, one to read
    // and one to write, to iterate over the input, when the element over the read
    // pointer is the same as the one n indexes back from the write pointer, it
    // skips it, otherwise, we know that we have less than n instances of that
    // element in the current output array, and we can add the current one.
    //
    // Time complexity: O(n) - We visit each value in the input once.
    // Space complexity: O(1) - We only use pointers.
    //
    // Runtime 0 ms Beats 100%
    // Memory 2.1 MB Beats 84.38%
    pub fn remove_duplicates(nums: &mut Vec<i32>) -> i32 {
        // The maximum number of duplicates allowed.
        const MAX_DUPLICATES: usize = 2;
        // Nothing to do.
        if nums.len() <= MAX_DUPLICATES {
            return nums.len() as i32;
        }
        // A write pointer.
        let mut w = MAX_DUPLICATES;
        for r in MAX_DUPLICATES..nums.len() {
            // Check if the value two positions back is the same.
            if nums[r] != nums[w - MAX_DUPLICATES] {
                nums[w] = nums[r];
                w += 1;
            }
        }
        w as i32
    }
}

// Tests.
fn main() {
    let tests = [
        (vec![1], 1),
        (vec![1, 1], 2),
        (vec![1, 1, 1, 2, 2, 3], 5),
        (vec![0, 0, 1, 1, 1, 1, 2, 3, 3], 7),
    ];
    for test in tests {
        assert_eq!(Solution::remove_duplicates(&mut Vec::from(test.0)), test.1);
    }
    println!("All tests passed!")
}
